### PR TITLE
Feature/knn transformer

### PR DESCRIPTION
**Pull request by milos-colic**

### CODE REVIEW

File: docs/source/models/spatial-knn.rst [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+observation O. The similarity is defined by a distance function d(O, S[i]). The most common distance function is the Euclidean.
+The nearest neighbour problem is a special case of the k-nearest neighbour problem, where k=1. The k-nearest neighbour
+problem is defined as finding k observations from a set S that are the most similar to the given observation O.
+
+.. figure:: ../images/nearest_neighbour.png
+   :figclass: doc-figure
+
+   Fig 1. Nearest neighbour problem
+
+We define spatial k-nearest neighbour problem as finding k observations from a set S that are the most similar to the
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:18]: This part is repeated from above:
"The k-nearest neighbour problem is defined as finding k observations from a set S that are the most similar to the given observation O."

File: docs/source/models/spatial-knn.rst [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+##############################
+
+The traditional definition of the nearest neighbour problem is defined as:
+"given a set S of points in a space M and a query point q in M, find the point in S that is closest to q".
+We have relaxed this definition by allowing the query point to be a set of points. The result of the nearest neighbour
+problem is a set of points from S that are closest to each query points. This in effect formulates a join condition
+where a point pair (Q[i], S[j]) is a match if st_distance(Q[i], S[j]) <= st_distance(Q[i], S[k]) where S[k] is the
+kth nearest neighbour of Q[i].
+
+The problem of finding the k-nearest neighbours can be formulated as a join condition where a point pair (Q[i], S[j]) is
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:35]: This phrase is incomplete/unclear

File: docs/source/models/spatial-knn.rst [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+Spatial KNN as an Iterative Transformation
+##########################################
+
+The problem of finding the k-nearest neighbours can be formulated as an iterative transformation. The transformation
+is defined as follows:
+* For each geometry in set Q generate a hex ring in grid index space
+* Generate match candidates in the hex ring
+* For each match candidate calculate the distance to the query geometry
+* For each geometry in set Q count the matches and stop if the count is equal to k
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:48]: Do you count the matches ordered by distance?

File: docs/source/models/spatial-knn.rst [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+* Continue with the next hex ring until max number of iterations is performed
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:49]: Do you always continue until max number of iterations, or do you stop if all Q have at least k neighbours?

File: docs/source/models/spatial-knn.rst (hunk 3 above)
Comment (edurdevic) [reply docs/source/models/spatial-knn.rst:48]: do you stop if count is equal to k, or at least k?

File: docs/source/models/spatial-knn.rst [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    >>> import mosaic as mos
+    >>> mos.enable_mosaic(spark, dbutils)
+    >>>
+    >>> buildings_df = spark.read.table("buildings")
+    >>> buildings_df = buildings_df.withColumn("left_id", monotonically_increasing_id())
+    >>> roads_df = spark.read.table("roads")
+    >>> roads_df = roads_df.withColumn("right_id", monotonically_increasing_id())
+    >>>
+    >>> from mosaic.models import SpatialKnn
+    >>> knn = SpatialKNN(roads_df)
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:75]: Is df_roads the Q or the S set?

File: docs/source/models/spatial-knn.rst [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    >>> knn.setLeftGeometryColumn("geometry")
+    >>> knn.setRightGeometryColumn("geometry")
+    >>> knn.setLeftIdColumn("left_id")
+    >>> knn.setRightIdColumn("right_id")
+    >>> knn.setK(5)
+    >>> knn.setMaxIterations(10)
+    >>> knn.setDistanceThreshold(1.0)
+    >>> knn.setCheckpointTablePrefix("checkpoint_table_knn")
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:83]: Do you need to set the Spark checkpoint directory before this?

File: docs/source/models/spatial-knn.rst [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+    >>> buildings_df = spark.read.table("buildings")
+    >>> buildings_df = buildings_df.withColumn("left_id", monotonically_increasing_id())
+    >>> roads_df = spark.read.table("roads")
+    >>> roads_df = roads_df.withColumn("right_id", monotonically_increasing_id())
+    >>>
+    >>> from mosaic.models import SpatialKnn
+    >>> knn = SpatialKNN(roads_df)
+    >>> knn.setLeftGeometryColumn("geometry")
+    >>> knn.setRightGeometryColumn("geometry")
+    >>> knn.setLeftIdColumn("left_id")
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:78]: Left/Right are not very clear.
Which one is S and which one is Q? Do you have some other candidate names? Q is usually the POI, right?

File: docs/source/models/spatial-knn.rst [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+
+Mosaic supports all indexing systems for this transformer.
+Please see :doc:`Spatial Indexing </api/spatial-indexing>` for supported indexing operations.
+
+Parameters
+##########
+
+The transformer has the following parameters:
+
+* leftGeometryColumn: the name of the column that contains the left geometry
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:147]: Left/right does not tell me which is Q and which is S

File: docs/source/models/spatial-knn.rst [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+* rightGeometryColumn: the name of the column that contains the right geometry
+* leftIdColumn: the name of the column that contains the left id
+* rightIdColumn: the name of the column that contains the right id
+* k: the number of neighbours to return
+* maxIterations: the maximum number of iterations to perform
+* distanceThreshold: the distance threshold to stop the iterations
+* checkpointTablePrefix: the prefix of the checkpoint table
+* indexResolution: the resolution of the index
+* approximate: whether to stop after max iterations (approximate = true) or to
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:156]: What is the default value here?

File: docs/source/models/spatial-knn.rst [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+  perform the finalisation step (approximate = false)
+
+If the approximate is set to true the transformer wont perform the finalisation step.
+The finalisation takes into account that grid index cells may be skewed at different
+locations and we cant ensure radial growth between iterations. That means that some
+of the neighbours in returned K set arent nearest neighbours. The finalisation step
+will take the distance between the neighbours and the target geometry and will generate
+a buffered geometry around the target geometry. The buffered geometry will be used to
+identify missed neighbours. The missed neighbours will be added to the K set and the
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:165]: Why using a buffering instead of adding an extra k-ring step?

File: docs/source/models/spatial-knn.rst [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+    >>>     knn.setRightIdColumn("right_id")
+    >>>     knn.setK(5)
+    >>>     knn.setMaxIterations(10)
+    >>>     knn.setDistanceThreshold(1.0)
+    >>>     knn.setCheckpointTablePrefix("checkpoint_table_knn")
+    >>>     knn.setIndexResolution(10)
+    >>>
+    >>>     neighbours = knn.transform(buildings_df)
+    >>>
+    >>>     mlflow.log_params(knn.getParams())
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:238]: Nice!

File: docs/source/models/spatial-knn.rst [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+.. raw:: html
+
+   </div>
+
+
+
+Shape Aware Hex Rings
+#####################
+
+When performing the iterations the transformer will use a hexagonal grid to identify
```
Comment (edurdevic) [inline docs/source/models/spatial-knn.rst:273]: Is it always hexagonal? What happens for rectangular grids? Is the k-ring in that case squared or does it approximate a circle?

File: pom.xml [hunk 13]
```diff
@@ -206,23 +218,25 @@
                 </executions>
             </plugin>
             <plugin>
-                <groupId>org.jacoco</groupId>
-                <artifactId>jacoco-maven-plugin</artifactId>
-                <version>0.8.8</version>
+                <groupId>org.scoverage</groupId>
```
Comment (edurdevic) [inline pom.xml:221]: Why was this changed?

File: python/mosaic/api/functions.py [hunk 14]
```diff
@@ -52,6 +52,15 @@
     "grid_tessellate",
     "grid_tessellateexplode",
 
+    "grid_cellkring",
```
Comment (edurdevic) [inline python/mosaic/api/functions.py:55]: Could you please add the docs for this functions?

File: python/mosaic/models/knn/spatial_knn.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
+from pyspark.sql import SparkSession, DataFrame, SQLContext
+from mosaic.utils import scala_utils
+
+class SpatialKNN:
+    def __init__(self, right_df):
```
Comment (edurdevic) [inline python/mosaic/models/knn/spatial_knn.py:5]: I would change here the same naming convention suggested above

File: python/mosaic/models/knn/spatial_knn.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+    def getParams(self):
+        params = self.model.getParams()
+        return scala_utils.scala_map_to_python(params)
+
+    def getMetrics(self):
+        metrics = self.model.getMetrics()
+        return scala_utils.scala_map_to_python(metrics)
+
+    def write(self):
```
Comment (edurdevic) [inline python/mosaic/models/knn/spatial_knn.py:76]: I would add a docstring that explains where it is writing to, and loading from (next method)

File: src/main/scala/com/databricks/labs/mosaic/core/Mosaic.scala [hunk 17]
```diff
@@ -1,21 +1,62 @@
...
+        geometry: MosaicGeometry,
+        resolution: Int,
+        keepCoreGeom: Boolean,
+        indexSystem: IndexSystem,
+        geometryAPI: GeometryAPI
+    ): Seq[MosaicChip] = {
+        GeometryTypeEnum.fromString(geometry.getGeometryType) match {
+            case POINT           => pointChip(geometry, resolution, keepCoreGeom, indexSystem)
+            case MULTIPOINT      => multiPointChips(geometry, resolution, keepCoreGeom, indexSystem)
+            case LINESTRING      => lineFill(geometry, resolution, indexSystem, geometryAPI)
```
Comment (edurdevic) [inline src/main/scala/com/databricks/labs/mosaic/core/Mosaic.scala:31]: Linestring andd Multilinestring both have another switch like this.
Is it worth refactoring this class to separate them out and rename all methods like lineChips, multil;ineChips, polygonChips?

File: src/main/scala/com/databricks/labs/mosaic/core/index/H3IndexSystem.scala [hunk 18]
```diff
@@ -32,7 +35,16 @@ object H3IndexSystem extends IndexSystem with Serializable {
...
       */
     override def getResolution(res: Any): Int = {
-        val resolution: Int = res.asInstanceOf[Int]
+        val resolution = (
+          Try(res.asInstanceOf[Int]),
+          Try(res.asInstanceOf[String]),
+          Try(res.asInstanceOf[UTF8String].toString)
+        ) match {
+            case (Success(i), _, _) => i
+            case (_, Success(s), _) => s.toInt
```
Comment (edurdevic) [inline src/main/scala/com/databricks/labs/mosaic/core/index/H3IndexSystem.scala:44]: The toInt should go inside the Try block, otherwise "abc" would not be caught as exception

File: src/main/scala/com/databricks/labs/mosaic/models/knn/HexRingNeighboursParams.scala [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+    val indexResolution: IntParam =
+        new IntParam(
+          this,
+          "indexResolution",
+          "Resolution for the index system to be used to represent geometries and to generate K rings as catchment are for neighbours."
+        )
+
+    def getLeftFeatureCol: String = $(leftFeatureCol)
+
+    def getLeftRowID: String = if (isDefined(leftRowID)) $(leftRowID) else "left_miid"
```
Comment (edurdevic) [inline src/main/scala/com/databricks/labs/mosaic/models/knn/HexRingNeighboursParams.scala:26]: what is "miid"?

File: src/main/scala/com/databricks/labs/mosaic/models/knn/SpatialKNN.scala [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+    override def earlyStoppingCheck(preDf: Dataset[_], postDf: Dataset[_]): Boolean = {
+        val preCount = preDf.count()
+        val postCount = postDf.count()
+        val newMatchesCount = matchesCheckpoint
+            .load()
+            .where("candidates_miid is not null")
+            .groupBy("landmarks_miid", "candidates_miid")
+            .count()
+            .count()
+        val condition = (preCount == postCount) && (matchCount == newMatchesCount) && (newMatchesCount > 0)
```
Comment (edurdevic) [inline src/main/scala/com/databricks/labs/mosaic/models/knn/SpatialKNN.scala:117]: `condition` -> `shouldStop` NIT

File: src/main/scala/com/databricks/labs/mosaic/models/knn/SpatialKNN.scala [hunk 21]
```diff
@@ -0,0 +1,326 @@
...
+        val newMatches = hexRingNeighboursTf.transform(dataset)
+        matches = matchesCheckpoint.append(newMatches)
+
+        val groupByCols = dataset.columns.filterNot(Seq("iteration", "match_radius", getLandmarksRowID).contains)
+        val result = matches
+            // Apply distance threshold but dont drop null matches
+            .where(coalesce(col(hexRingNeighboursTf.distanceCol) <= getDistanceThreshold, lit(true)))
+            .groupBy(getLandmarksRowID, groupByCols: _*)
+            .agg(
+              sum(col(getCandidatesRowID).isNotNull.cast("int")).alias("match_count"),
```
Comment (edurdevic) [inline src/main/scala/com/databricks/labs/mosaic/models/knn/SpatialKNN.scala:146]: Should not getCandidatesRowID be a long type?

Comment (edurdevic) [reply src/main/scala/com/databricks/labs/mosaic/models/knn/SpatialKNN.scala:146]: And why summing the rowID?

File: src/main/scala/com/databricks/labs/mosaic/models/knn/SpatialKNNParams.scala [hunk 22]
```diff
@@ -0,0 +1,80 @@
+package com.databricks.labs.mosaic.models.knn
+
+import com.databricks.labs.mosaic.models.core.IterativeTransformerParams
+import org.apache.spark.ml.param._
+
+trait SpatialKNNParams extends Params with IterativeTransformerParams {
+
+    val useTableCheckpoint: BooleanParam =
+        new BooleanParam(this, "useTableCheckpoint", "Use a table checkpoint to store intermediate results.")
```
Comment (edurdevic) [inline src/main/scala/com/databricks/labs/mosaic/models/knn/SpatialKNNParams.scala:9]: Where do you set the table name when this is true?

File: src/main/scala/com/databricks/labs/mosaic/models/util/CheckpointManager.scala [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+import org.apache.spark.sql.Dataset
+
+/**
+  * CheckpointManager is a utility class to manage checkpointing of the mosaic
+  * model.
+  *
+  * @param checkpoint
+  *   Either DeltaFile or DeltaTable checkpoint instance.
+  */
+case class CheckpointManager(checkpoint: DeltaCheckpoint) {
```
Comment (edurdevic) [inline src/main/scala/com/databricks/labs/mosaic/models/util/CheckpointManager.scala:12]: Why do we need a dedicated class for checkpoointing? What is the difference with Spark checkpoint/localCheckpoint?

**Review by edurdevic — COMMENTED** [review]
I reviewed the code.
I still need to review the tests

**Comment by milos-colic** [comment]
@edurdevic I have fixed all your comments. pushing the commit now.

**Review by edurdevic — APPROVED** [review]
Awesome work @milos-colic!

**Comment by milos-colic** [comment]
@edurdevic just pushed the last fix you highlighted - after the build is successful we can merge

**Review by edurdevic — APPROVED** [review]
